src/io/fastx.rs

```diff
@@ -0,0 +1,442 @@
+            Record::FASTQ(x) => x.check(),
+        }
+    }
+
+    pub fn desc(&self) -> Option<&str> {
+        match self {
+            Record::FASTA(x) => x.desc(),
+            Record::FASTQ(x) => x.desc(),
+        }
+    }
+
+    pub fn seq(&self) -> TextSlice<'_> {
+        match self {
+            Record::FASTA(x) => x.seq(),
+            Record::FASTQ(x) => x.seq(),
+        }
+    }
+
+    pub fn qual(&self) -> Vec<u8> {
+        match self {
+            Record::FASTA(x) => vec![BASE_QUAL; x.seq().len()],
+            Record::FASTQ(x) => x.qual().into(),
+        }
+    }
+
+    pub fn set_base_qual(&mut self, qual: &[u8]) {
+        let new_record = fastq::Record::with_attrs(self.id(), self.desc(), self.seq(), qual);
+        mem::replace(self, Record::FASTQ(new_record));
+    }
+}
+
+impl SequenceRead for Record {
+    fn name(&self) -> &[u8] {
+        self.id().as_bytes()
+    }
+
+    fn base(&self, i: usize) -> u8 {
+        self.seq()[i]
+    }
+
+    fn base_qual(&self, i: usize) -> u8 {
+        match self {
+            Record::FASTQ(x) => x.base_qual(i),
+            Record::FASTA(_x) => BASE_QUAL,
+        }
+    }
+
+    fn len(&self) -> usize {
+        self.seq().len()
+    }
+}
+
+impl<R> fasta::FastaRead for Reader<R>
+where
+    R: io::Read,
+{
+    fn read(&mut self, record: &mut fasta::Record) -> io::Result<()> {
+        if let Reader::Uninitialized(_) = self {
+            self.init_reader()?;
+        };
+
+        match self {
+            Reader::FASTA(x) => x.read(record),
+            Reader::FASTQ(x) => {
+                let mut fq_record = fastq::Record::new();
+                x.read(&mut fq_record)?;
+                mem::replace(
+                    record,
+                    fasta::Record::with_attrs(fq_record.id(), fq_record.desc(), fq_record.seq()),
+                );
+                Ok(())
+            }
+            Reader::Uninitialized(_) => unreachable!(),
+        }
+    }
+}
+
+impl<R> fastq::FastqRead for Reader<R>
+where
+    R: io::Read,
+{
+    fn read(&mut self, record: &mut fastq::Record) -> io::Result<()> {
+        if let Reader::Uninitialized(_) = self {
+            self.init_reader()?;
+        };
+
+        match self {
+            Reader::FASTA(x) => {
+                let mut fa_record = fasta::Record::new();
+                x.read(&mut fa_record)?;
+                mem::replace(
+                    record,
+                    fastq::Record::with_attrs(
+                        fa_record.id(),
+                        fa_record.desc(),
+                        fa_record.seq(),
+                        &vec![BASE_QUAL; fa_record.seq().len()],
+                    ),
+                );
+                Ok(())
+            }
+            Reader::FASTQ(x) => x.read(record),
+            Reader::Uninitialized(_) => unreachable!(),
+        }
+    }
+}
+
+impl<R: io::Read> Reader<R> {
+    pub fn new(reader: R) -> Self {
+        Reader::Uninitialized(Some(reader))
```

Comment:
Delaying the initialization to when `read` is called, and `init_reader` then changes the `Reader::uninitialized` into the proper `fasta::Reader` or `fastq::Reader`